stacks/the-stack-abstract-data-type.md

```diff
@@ -1,7 +1,31 @@
 The Stack Abstract Data Type
 ============================
 
-The stack abstract data type is an ordered collection of items where items are added to and removed from the end called the “top.” Stacks are ordered LIFO. The interface for a stack is:
+An *abstract data type*, sometimes abbreviated *ADT*, is a logical
+description of how we view the data and the operations that are allowed
+without regard to how they will be implemented. This means that we are
+concerned only with what the data is representing and not with how it
+will eventually be constructed. By providing this level of abstraction,
+we are creating an *encapsulation* around the data. The idea is that by
+encapsulating the details of the implementation, we are hiding them from
+the user’s view. This is called *information hiding*.
+
+The implementation of an abstract data type, often referred to as a
+*data structure*, will require that we provide a physical view of the
+data using some collection of programming constructs and primitive data
+types. The separation of these two perspectives will allow us to define
+the complex data models for our problems without giving any indication
+as to the details of how the model will actually be built. This provides
+an *implementation-independent* view of the data. Since there will
+usually be many different ways to implement an abstract data type, this
+implementation independence allows the programmer to switch the details
+of the implementation without changing the way the user of the data
+interacts with it. The user can remain focused on the problem-solving
+process.
```

Comment:
I like introducing the definition of ADT later outside of the intro

I'm not sure that describing it right before a stack makes sense.  Ruby/Python/JS all have just one obvious correct way to implement a stack... so what's the benefit to knowing what an ADT is, exactly?

Compare that to the ways you could implement checking whether a node exists in a tree. All of the sudden the benefit of encapsulating the implementation of the ADT is useful.